java/com/google/re2j/Inst.java

```diff
@@ -11,41 +11,44 @@
  * A single instruction in the regular expression virtual machine.
  * @see http://swtch.com/~rsc/regexp/regexp2.html
  */
-class Inst {
+final class Inst {
+  
+  public static final int ALT = 1;
+  public static final int ALT_MATCH = 2;
+  public static final int CAPTURE = 3;
+  public static final int EMPTY_WIDTH = 4;
+  public static final int FAIL = 5;
+  public static final int MATCH = 6;
+  public static final int NOP = 7;
+  public static final int RUNE = 8;
+  public static final int RUNE1 = 9;
+  public static final int RUNE_ANY = 10;
+  public static final int RUNE_ANY_NOT_NL = 11;
 
-  enum Op {
-    ALT,
-    ALT_MATCH,
-    CAPTURE,
-    EMPTY_WIDTH,
-    FAIL,
-    MATCH,
-    NOP,
-    RUNE,
-    RUNE1,
-    RUNE_ANY,
-    RUNE_ANY_NOT_NL,
-  }
-
-  Op op;
-  int out;  // all but MATCH, FAIL
-  int arg;  // ALT, ALT_MATCH, CAPTURE, EMPTY_WIDTH
-  int[] runes;  // length==1 => exact match
+  final int op;
```

Comment:
It might be worth doing a separate experiment to see if the compactness gain from putting each of these four int fields into a byte makes the Prog faster.

Also, if runes is commonly of length zero or 1, consider special-casing it.

Reply:
might try that see if that change anything. because the object tend to be padded for alignement might make no difference though...

Reply:
Even with padding, this change replaces 16 bytes by 4, which is at least a word saved.